_posts/2020-05-18-mirror-kafka-eventhub.md

```diff
@@ -0,0 +1,225 @@
+Anyway, it's worth mentioning that the Kubernetes cluster, where Apache Kafka runs, could be an AKS (Azure Kubernetes Service) one so as part of the same Microsoft Cloud infrastracture.
+
+The first step is storing the Event Hub authentication info into a Kubernetes `Secret` in order to be referenced in the Strimzi `KafkaMirrorMaker` custom resource for enabling the producer part to connect to the Event Hub itself.
+The following snippet shows such a `Secret` where the `$ConnectionString` hasn't to be changed while the endpoint has to be customized with the actual Event Hub connection string; save this `Secret` in a file named `eventhubs-secret.yaml`
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: eventhubssecret
+type: Opaque
+stringData:
+  eventhubspassword: Endpoint=sb://<eventhubs-namespace>.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=<access-key>
+```
+
+Create the `Secret` on the Kubernetes cluster.
+
+```shell
+kubectl apply -f eventhubs-secret.yaml -n kafka
+```
+
+The Kafka Mirror Maker instance is deployed via the Strimzi cluster operator through a corresponding `KafkaMirrorMaker` resource as the following.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaMirrorMaker
+metadata:
+  name: my-mirror-maker
+spec:
+  version: 2.4.0
+  replicas: 1
+  consumer:
+    bootstrapServers: my-cluster-kafka-bootstrap:9092
+    groupId: my-source-group-id
+  producer:
+    bootstrapServers: <eventhubs-namespace>.servicebus.windows.net:9093
+    authentication:
+      type: plain
+      username: $ConnectionString
+      passwordSecret:
+        secretName: eventhubssecret
+        password: eventhubspassword
+    tls:
+      trustedCertificates: []
+  whitelist: ".*"
+```
+
+The `consumer` part just connects to the local Apache Kafka cluster running on the same Kubernetes instance.
+
+The `producer` part connects to the Event Hub namespace, through TLS and PLAIN authentication using `$ConnectionString` as username and the connection string as password provided in the already created `eventhubssecret`.
+The `tls` section is used because Event Hub connection [needs SSL](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview?WT.mc_id=devto-blog-abhishgu#security-and-authentication) with `SASL_SSL` as security protocol
+
+Create the `KafkaMirrorMaker` on the Kubernetes cluster saving the above resource into a `kafka-mirror-maker.yaml` file.
+The Strimzi cluster operator takes care of it deploying Kafka Mirror Maker using the above configuration.
+
+```shell
+kubectl apply -f kafka-mirror-maker.yaml -n kafka
+```
+
+### Produce, mirror and trigger!
+
+The application consuming the mirrored messages from Event Hub is a simple Azure Functions application which just gets and logs the message that is enough for the purpose of this demo.
```

Comment:
For the purpose of this demo, the application consuming the mirrored messages from Event Hub is a simple Azure Functions application which just gets and logs the message.